### PR TITLE
Update major release of the the python packages on the jupyterlab notebooks

### PR DESCRIPTION
**Pull request by atheo89**
Related to: https://issues.redhat.com/browse/RHOAIENG-11080 

**NOTE**: Please do not squash the commits 

## Description

This PR proposes upgrading the JupyterLab package from 3.6.7 to 4.2.4. In addition to the JupyterLab upgrade, other dependent packages have also been updated to versions compatible with the new release. Moreover, it removes `odh-elyra` bits till elyra be fully supported.

List of the images that affected by this PR (ubi9-python-3.11 flavor)
- base/ubi9-python-3.11
- jupyter/minimal
- jupyter/datascience
- jupyter/tensorflow
- jupyter/pytorch
- jupyter/trustyai
- jupyter/rocm/tensorflow
- jupyter/rocm/pytorch

**Note:** This PR does not include odh-elyra package.  
For the phase one, we excluded elyra installation, till this Jira ticket iget resolved. https://issues.redhat.com/browse/RHOAIENG-10143

**NOTE**: Please do not squash the commits 

## How Has This Been Tested?
Check status of the image builds from gha. 

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by atheo89** [comment]
The notebooks from data science and after are failing in this step in the dockerfile: https://github.com/opendatahub-io/notebooks/blob/main/jupyter/datascience/ubi9-python-3.11/Dockerfile#L55 

More specifically, this part https://github.com/opendatahub-io/notebooks/blob/main/jupyter/datascience/ubi9-python-3.11/Dockerfile#L55-L71 is related to elyra piplines. Since the [odh-elyra](https://github.com/atheo89/notebooks/blob/RHOAIENG-11080/jupyter/datascience/ubi9-python-3.11/Pipfile#L31) package is not included (for now) on the related piplock file to install elyra, does not work properly.

**NOTE: I will comment this part temporarily to check the functionality, and i will revert it back before the merge**

File: base/ubi9-python-3.11/Pipfile [hunk 1]
```diff
@@ -7,8 +7,8 @@ verify_ssl = true
 
 [packages]
 # Base packages
-wheel = "~=0.43.0"
-setuptools = "~=69.2.0"
+wheel = "~=0.44.0"
+setuptools = "~=73.0.0"
```
Comment (jiridanek) [inline base/ubi9-python-3.11/Pipfile:11]: meanwhile, 74.0.0 came out ;P I guess let's ignore that and stay with 73

https://pypi.org/project/setuptools/

File: jupyter/datascience/ubi9-python-3.11/Pipfile [hunk 2]
```diff
@@ -7,44 +7,50 @@ name = "pypi"
 
 [packages]
 # Datascience and useful extensions
-boto3 = "~=1.34.50"
+boto3 = "~=1.35.5"
```
Comment (jiridanek) [inline jupyter/datascience/ubi9-python-3.11/Pipfile:10]: crazy, now 1.35.8 is the latest; they release this all the time! https://pypi.org/project/boto3/

**Review by jiridanek — COMMENTED** [review]
/lgtm

what's updated looks good, and I don't think we need to be catching up to the very latest versions all the time

**Comment by jiridanek** [comment]
> Build Notebooks (pr) / build (cuda-jupyter-tensorflow-ubi9-python-3.11) / build (pull_request) Failing after 15m

that's a legit test fail

File: jupyter/datascience/ubi9-python-3.11/Pipfile (hunk 2 above)
Comment (atheo89) [reply jupyter/datascience/ubi9-python-3.11/Pipfile:10]: true, boto3 do release almost every day!

**Comment by atheo89** [comment]
Thanks for the review! I will bring back the commit that exclude elyra bits. 
So expect lot of fireworks :P

**Comment by atheo89** [comment]
Please note that, as discussed, we will temporarily remove **Elyra** components until everything is ready on elyra side. Elyra will be excluded in version 2.14, with plans to reintroduce it in version 2.15
cc: @harshad16 @adrielparedes

File: jupyter/datascience/ubi9-python-3.11/Pipfile [hunk 3]
```diff
@@ -7,44 +7,50 @@ name = "pypi"
...
+matplotlib = "~=3.9.2"
+numpy = "~=2.1.0"
+pandas = "~=2.2.2"
+plotly = "~=5.23.0"
+scikit-learn = "~=1.5.1"
+scipy = "~=1.14.1"
+skl2onnx = "~=1.17.0"
+onnxconverter-common = "~=1.14.0" # Required for skl2onnx, as upgraded version is not compatible with protobuf
 codeflare-sdk = "~=0.18.0"
 memray = "~=1.10.0"  # any version above this would cause issue with rich package
```
Comment (harshad16) [inline jupyter/datascience/ubi9-python-3.11/Pipfile:21]: Can we upgrade this as well,
and as the memray was included due to codeflare-sdk version, 
can you try and see if removing it gives any issue. 
i believe we can remove it from direct list.

File: jupyter/datascience/ubi9-python-3.11/Pipfile [hunk 4]
```diff
@@ -7,44 +7,50 @@ name = "pypi"
...
+nbgitpuller = "~=1.2.1"
+
 # pycodestyle is dependency of below packages
 # and to achieve compatible of pycodestyle with python-lsp-server[all]
 # pinned the below packages
-autopep8 = "~=2.0.4"
-flake8 = "~=7.0.0"
+autopep8 = "~=2.3.1"
+flake8 = "~=7.1.1"
+
```
Comment (harshad16) [inline jupyter/datascience/ubi9-python-3.11/Pipfile:50]: I believe these can be removed from direct list , as odh-elyra is removed.
they came from there.

**Review by harshad16 — CHANGES_REQUESTED** [review]
This is an awesome work.
some suggestions.

File: jupyter/datascience/ubi9-python-3.11/Pipfile (hunk 3 above)
Comment (atheo89) [reply jupyter/datascience/ubi9-python-3.11/Pipfile]: Hey I removed memray package, you may inspect it here https://github.com/opendatahub-io/notebooks/pull/687/commits/e0769d2439a4c4b23c780cbe35db088f13043e7f

**Comment by atheo89** [comment]
>@atheo89 , This PR is merged, https://github.com/opendatahub-io/notebooks/pull/678
can you please rebase and take those changes in. 

Hey Harshad, I rebased the PR, but i kept the changes generated from this one, since the latest version of `codeflare-sdk` is not  [compatible](https://github.com/project-codeflare/codeflare-sdk/blob/main/pyproject.toml#L35) with setuptools >70, so we need to pindown to 0.18. 
I have opened an issue tracker on jira https://issues.redhat.com/browse/RHOAIENG-12507

**Review by harshad16 — APPROVED** [review]
/lgtm
/approve

Thanks for this , this is a good start.

**Comment by harshad16** [comment]
/override ci/prow/notebook-jupyter-intel-pyt-ubi9-python-3-11-pr-image-mirror
/override ci/prow/runtimes-ubi9-e2e-tests
/override ci/prow/codeserver-notebook-e2e-tests
/override ci/prow/notebooks-ubi9-e2e-tests
/override ci/prow/rocm-notebooks-e2e-tests
/override ci/prow/images

**Pull request merged**